File: src/aoc23/day01.clj
(ns aoc23.day01
  (:require [clojure.string :as str]
            [clojure.edn :as edn]
            [aoc23.util :as util]))

(def testf "data/day01-test.txt")
(def test2f "data/day01-test2.txt")
(def inputf "data/day01-input.txt")

(defn get-numbers
  "Extract each digit from the string"
  [s]
  (->> (str/split s #"")
       (map edn/read-string)
       (filter number?)))

(defn get-numbers-words
  "Parse the string into a list of digits, stored as either numbers or words"
  [s]
  (let [m {"one" 1 "two" 2 "three" 3 "four" 4 "five" 5 "six" 6 "seven" 7 "eight" 8 "nine" 9}
        pattern #"(?=(one|two|three|four|five|six|seven|eight|nine|\d))"
        numbers (map last (re-seq pattern s))]
    (map #(or (get m %) (edn/read-string %)) numbers)))

(defn make-number
  "Make a 2-digit number from the digits in the collection"
  [coll]
  (+ (* 10 (first coll)) (last coll)))

(defn part1
  [f]
  (let [data (util/import-data f)]
    (->> data
         (map get-numbers)
         (map make-number)
         (apply +))))

(defn part2
  [f]
  (let [data (util/import-data f)]
    (->> data
         (map get-numbers-words)
         (map make-number)
         (apply +))))

(comment 
  (assert (= 142 (part1 testf))))
  (assert (= 281 (part2 test2f)))

File: src/day01.jl
# day01.jl

module Day01

using Chain, Test, Main.Util

"""Find the aggregate of all digits in a string"""
function find_numbers(str)
  digits = [m.match for m in eachmatch(r"\d", str)]
  parse(Int, first(digits) * last(digits))
end

function convert_match(str)
  m = Dict("one" => "1", "two" => "2", "three" => "3", "four" => "4", "five" => "5",
           "six" => "6", "seven" => "7", "eight" => "8", "nine" => "9")
  if str ∈ keys(m)
    return m[str]
  else
    return str
  end
end

"""Find numbers as digits or words"""
function find_numbers2(str)
  digits = r"one|two|three|four|five|six|seven|eight|nine|\d"
  numbers = [m.match |> convert_match for m in eachmatch(digits, str, overlap = true)]
  parse(Int, first(numbers) * last(numbers))
end

function part1(f)
  lines = Util.read_data(f)
  @chain lines begin
    filter(x -> length(x)>=1, _)
    map(find_numbers, _)
    sum
  end
end

function part2(f)
  lines = Util.read_data(f)
  @chain lines begin
    filter(x -> length(x)>=1, _)
    map(find_numbers2, _)
    sum
  end
end

@test part1("data/day01-test.txt") == 142
@test part2("data/day01-test.txt") == 142
@test part2("data/day01-test2.txt") == 281

end # module

# The End